File: Examples/ObjectiveC/MissingIconsExample.m
#import "MissingIconsExample.h"
@import Mapbox;

NSString *const MBXExampleMissingIcons = @"MissingIconsExample";

@interface MissingIconsExample ()<MGLMapViewDelegate>

@property (strong, nonatomic) MGLMapView *mapView;

@end

@implementation MissingIconsExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // TODO: This is a test case, it should be changed to fulfill an ios example spec.
    self.mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds styleURL:[MGLStyle streetsStyleURL]];
    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(0, 0) zoomLevel:1 animated:NO];
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    self.mapView.delegate = self;
    [self.view addSubview:self.mapView];
    
    // Create a button to load the faulty style.
    UIButton *button = [[UIButton alloc] init];
    button.translatesAutoresizingMaskIntoConstraints = NO;
    button.layer.cornerRadius = 15;
    button.backgroundColor = [UIColor whiteColor];
    [button setTitle:@"Load faulty style" forState:UIControlStateNormal];
    [button setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    button.titleEdgeInsets = UIEdgeInsetsMake(5, 5, 5, 5);
    button.titleLabel.adjustsFontSizeToFitWidth = YES;
    [button addTarget:self action:@selector(loadFaultyStyle:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:button];
    
    if (@available(iOS 11.0, *)) {
        [NSLayoutConstraint activateConstraints:@[[button.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:20],
                                                  [button.leftAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leftAnchor constant:20]
                                                  ]];
    } else {
        [NSLayoutConstraint activateConstraints:@[[button.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:80],
                                                  [button.leftAnchor constraintEqualToAnchor:self.view.leftAnchor constant:20]
                                                  ]];
    }
 
}

- (void)loadFaultyStyle:(id) sender {
    NSURL *customStyleJSON = [[NSBundle mainBundle] URLForResource:@"missing_icon" withExtension:@"json"];
    [self.mapView setStyleURL:customStyleJSON];
}

- (UIImage *)mapView:(MGLMapView *)mapView didFailToLoadImage:(NSString *)imageName {
    if (![imageName isEqualToString:@"skip-this-missing-icon"]) {
        UIImage *backupImage = [UIImage imageNamed:@"mapbox"];
        return backupImage;
    }
    return nil;
}

@end

File: Examples/Swift/MissingIconsExample.swift
import UIKit
import Mapbox

@objc(MissingIconsExample_Swift)

class MissingIconsExample: UIViewController {

    let mapView = MGLMapView()

    override func viewDidLoad() {
        super.viewDidLoad()

        // TODO: This is a test case, it should be changed to fulfill an ios example spec.
        mapView.styleURL = MGLStyle.streetsStyleURL
        mapView.centerCoordinate = CLLocationCoordinate2D(latitude: 0, longitude: 0)
        mapView.frame = CGRect(x: 0, y: 0, width: view.bounds.width, height: view.bounds.height)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]

        mapView.delegate = self
        view.addSubview(mapView)

        // Create a button to load the faulty style.
        let button = UIButton()
        button.translatesAutoresizingMaskIntoConstraints = false
        button.layer.cornerRadius = 15
        button.backgroundColor = UIColor.white
        button.setTitle("Load faulty style", for: .normal)
        button.setTitleColor(UIColor.black, for: .normal)
        button.titleEdgeInsets = UIEdgeInsets(top: 5, left: 5, bottom: 5, right: 5)
        button.titleLabel?.adjustsFontSizeToFitWidth = true
        button.addTarget(self, action: #selector(loadFaultyStyle(sender:)), for: .touchUpInside)
        view.addSubview(button)

        if #available(iOS 11.0, *) {
            NSLayoutConstraint.activate([button.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
                                         button.leftAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leftAnchor, constant: 20)])
        } else {
            // This code should be enabled once we require iOS 9.0 as minimum target.
//            NSLayoutConstraint.activate([button.topAnchor.constraint(equalTo: view.topAnchor, constant: 80),
//                                         button.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 20)])
        }
    }

    @objc func loadFaultyStyle(sender: Any?) {
        let customStyleJSON = Bundle.main.url(forResource: "missing_icon", withExtension: "json")
        mapView.styleURL = customStyleJSON
    }

}

extension MissingIconsExample: MGLMapViewDelegate {

    func mapView(_ mapView: MGLMapView, didFailToLoadImage imageName: String) -> UIImage? {

        if (!(imageName == "skip-this-missing-icon")) {
            let backupImage = UIImage(named: "mapbox")
            return backupImage
        }
        return nil
    }

}
